File: packages/ui/src/client-plugins/bundle.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { CLIENT_SDK_VERSION, type ClientBundle, type ClientDescriptor } from '@maka-agent/plugin-sdk/client';

type Factory = ClientBundle['factory'];
interface Pending {
  readonly id: string;
  factory?: Factory;
  error?: Error;
}
const pendingDocuments = new WeakMap<Document, Map<string, Pending>>();

declare global {
  interface Window { __MakaClientBundle__?: (bundle: ClientBundle) => void }
}

/** Classic scripts have removable URLs and no permanent browser ESM module-map entry. */
export async function loadClientBundle(
  descriptor: ClientDescriptor,
  source: string,
  document: Document,
  signal: AbortSignal,
): Promise<Factory> {
  signal.throwIfAborted();
  if (descriptor.sdkVersion !== CLIENT_SDK_VERSION) throw new Error('Incompatible Client SDK');
  const bytes = new TextEncoder().encode(source);
  if (bytes.length !== descriptor.totalBytes) throw new Error('Client bundle length mismatch');
  const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', bytes));
  const hash = 'sha256-' + Array.from(digest, (byte) => byte.toString(16).padStart(2, '0')).join('');
  if (hash !== descriptor.clientDigest) throw new Error('Client bundle digest mismatch');
  signal.throwIfAborted();
  const window = document.defaultView;
  if (!window) throw new Error('Client document is closed');
  let pending = pendingDocuments.get(document);
  if (!pending) {
    pending = new Map();
    pendingDocuments.set(document, pending);
    const loads = pending;
    window.__MakaClientBundle__ = (bundle) => {
      const script = document.currentScript as HTMLScriptElement | null;
      const request = script ? loads.get(script.src) : undefined;
      if (!request) throw new Error('Client bundle registered outside its loading document');
      if (request.factory || bundle.id !== request.id || typeof bundle.factory !== 'function') {
        request.error = new Error('Invalid or duplicate Client bundle registration');
        throw request.error;
      }
      request.factory = bundle.factory;
    };
  }
  const url = URL.createObjectURL(new Blob([bytes], { type: 'text/javascript' }));
  const script = document.createElement('script');
  const request: Pending = { id: descriptor.extensionId };
  pending.set(url, request);
  script.src = url;
  const scriptError = (event: ErrorEvent) => {
    if (event.filename === url) request.error = new Error(event.message || 'Client bundle threw');
  };
  window.addEventListener('error', scriptError);
  let abort: (() => void) | undefined;
  try {
    await new Promise<void>((resolve, reject) => {
      abort = () => reject(signal.reason);
      const finish = (error?: unknown) => {
        error ? reject(error) : resolve();
      };
      script.onload = () => finish(request.error);
      script.onerror = () => finish(new Error('Client bundle script failed'));
      signal.addEventListener('abort', abort, { once: true });
      document.head.append(script);
      if (signal.aborted) abort();
    });
    signal.throwIfAborted();
    if (!request.factory) throw new Error('Client bundle did not register a factory');
    return request.factory;
  } finally {
    if (abort) signal.removeEventListener('abort', abort);
    window.removeEventListener('error', scriptError);
    script.onload = null;
    script.onerror = null;
    script.remove();
    pending.delete(url);
    URL.revokeObjectURL(url);
  }
}

File: crates/plugins/src/package/bundle.rs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

use super::{MAX_FILE_BYTES, MAX_FILES, MAX_PACKAGE_BYTES, Package};
use crate::Error;
use base64::{Engine, engine::general_purpose::STANDARD};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::BTreeMap;

#[derive(Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct Bundle {
    schema_version: u8,
    digest: String,
    files: Vec<BundleFile>,
}

#[derive(Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct BundleFile {
    path: String,
    sha256: String,
    content: String,
}

impl Package {
    pub fn from_bundle(bytes: &[u8]) -> Result<Self, Error> {
        if bytes.len() > MAX_PACKAGE_BYTES * 2 {
            return Err(Error::Invalid("encoded bundle exceeds 32 MiB".into()));
        }
        let bundle: Bundle = serde_json::from_slice(bytes)
            .map_err(|error| Error::Invalid(format!("bundle: {error}")))?;
        if bundle.schema_version != 1 || bundle.files.len() > MAX_FILES {
            return Err(Error::Invalid(
                "unsupported bundle schema or file count".into(),
            ));
        }
        let mut files = BTreeMap::new();
        let mut total = 0;
        for file in bundle.files {
            super::validate_path(&file.path)?;
            if file.content.len() > MAX_FILE_BYTES.div_ceil(3) * 4 {
                return Err(Error::Invalid("encoded bundle file exceeds limit".into()));
            }
            let bytes = STANDARD
                .decode(file.content)
                .map_err(|_| Error::Invalid("invalid bundle base64".into()))?;
            total += bytes.len();
            if bytes.len() > MAX_FILE_BYTES || total > MAX_PACKAGE_BYTES {
                return Err(Error::Invalid("bundle exceeds byte limits".into()));
            }
            if format!("{:x}", Sha256::digest(&bytes)) != file.sha256 {
                return Err(Error::Invalid("bundle file digest mismatch".into()));
            }
            if files.insert(file.path, bytes).is_some() {
                return Err(Error::Invalid(
                    "bundle contains duplicate file paths".into(),
                ));
            }
        }
        let package = Self::new(files)?;
        if package.digest() != bundle.digest {
            return Err(Error::Invalid("bundle package digest mismatch".into()));
        }
        Ok(package)
    }

    pub fn to_bundle(&self) -> Vec<u8> {
        let bundle = Bundle {
            schema_version: 1,
            digest: self.digest().into(),
            files: self
                .files()
                .iter()
                .map(|(path, bytes)| BundleFile {
                    path: path.clone(),
                    sha256: format!("{:x}", Sha256::digest(bytes)),
                    content: STANDARD.encode(bytes),
                })
                .collect(),
        };
        serde_json::to_vec(&bundle).expect("bundle contains only serializable strings")
    }
}
